Add runtime introductory offer checks to email upsell in domain flow

daledupreez opened the pull request:
#### Changes proposed in this Pull Request

This PR tackles a few items:
* It defines a new `ProductIntroductoryOffer` type and documents it as an optional property on the `ProductListItem` object returned from the `getProductBySlug()` and related selectors
* It adds a new `getProductIntroductoryOffer()` selector to return the introductory offer defined for a product slug
* It updates the `EmailProvidersComparison` component to rely on the presence of an introductory offer for the Google Workspace product instead of a feature flag

The combined goal of these changes is to make it possible for this page to dynamically detect when an introductory offer is available for Google Workspace, and then show the appropriate pricing details. This will mean this page will automatically switch to the correct pricing details when we make free trials for Google Workspace available in production.

#### Testing instructions

* Run this branch locally or via the [calypso.live branch](https://github.com/Automattic/wp-calypso/pull/61418#issuecomment-1049152298)
* Ensure you have store sandbox disabled
* Navigate to Upgrades -> Domains, and click on the "Add a domain" button in the top right corner
* Select "Search for a domain" from the dropdown
* Search for a domain name, and select a paid domain name from the list of suggestions
* Verify that you see the standard pricing for Google Workspace in the next page, without an indication of the free trial
* Click on one of the "Skip" options on the page
* Remove all products from your cart in the checkout page
* Enable store sandbox mode
* Ensure that the Calypso data store is empty by opening the developer tools, and ensuring that the IndexedDB data for Calypso is cleared (this is under the Application tab in Chrome, under `IndexedDB` -> `calypso - {hostname}` -> `calypso_store`, and you can clear the contents by right clicking on "calypso_store" and selecting the "Clear" option
* Navigate to Upgrades -> Domains and follow the same steps to add a domain to your cart
* Verify that when you get to the email upsell page, we now show the details for the Google Workspace 1 month free trial -- this is expected because we have enabled the free trial in the store sandbox

Related to #61324

daledupreez committed:
Add getProductIntroductoryOffer() selector and ProductIntroductoryOffer type

daledupreez committed:
Add checks for introductory offer and remove check for feature flag

AllTerrainDeveloper commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -481,13 +484,13 @@ class EmailProvidersComparison extends Component {
-		const discount = this.getAvailableDiscountForGoogle(
+		const discount = this.getAvailableDiscountForGoogle( {
 			currencyCode,
 			isEligibleForFreeTrial,
 			gSuiteProduct,
 			productIsDiscounted,
-			standardPrice
-		);
+			standardPrice,
+		} );
```
Not a change request, but curiosity! What's the benefit of encapsulated parameters?

stephanethomas replied:
Same question 😀.

daledupreez replied:
My (admittedly strong) personal preference is to use encapsulated parameters in any situation where I expect the parameter list to grow beyond 1 parameter. My thinking is that this makes it easier to add parameters over time in a way that is seamless, and generally requires fewer changes to callers (especially when you get to adding a second and later parameters). I also like it because it forces me to treat the function more like a component that receives props - it helps to make more dependencies injected/supplied and makes it easy to refactor the logic into a React component if needed.

AllTerrainDeveloper commented on client/state/products-list/selectors/get-product-introductory-offer.ts:
```diff
@@ -0,0 +1,18 @@
+import type { ProductIntroductoryOffer } from './get-products-list';
+import type { AppState } from 'calypso/types';
+
+/**
+ * Retrieves an introductory offer for the product with the specified slug.
+ *
+ * @param {object} state - global state tree
+ * @param {string} productSlug - internal product slug, eg 'jetpack_premium'
+ * @returns {import('./get-products-list').ProductIntroductoryOffer|null} - introductory offer details for the product, or null if no offer exists
+ */
+
```
Does this docblock makes sense on a TS? The function definition seems to me already self descriptive enough. Plus also I think there is an extra space that we usually don't include between the docblock and the function.

daledupreez replied:
I was mostly replicating the approach for `getProductBySlug()`/`get-product-by-slug.ts` in the same directory.

I'll remove the extra whitespace.

AllTerrainDeveloper reviewed (commented):
Minor comment, not mandatory at all :) 

I'm testing it in a while and post the EOT.

AllTerrainDeveloper reviewed (approved):
Code with the minor comment looks good to me, plus also the tests are successful.

### EOT

Not showing any offer while not being sandboxed:
![image](https://user-images.githubusercontent.com/5689927/155486239-d62955b3-cba7-4569-b149-1d4fba11fb5e.png)

Showing the free month while being sandboxed:
![image](https://user-images.githubusercontent.com/5689927/155486537-d61b5365-8c00-4df8-919c-2fbd335d7a82.png)

stephanethomas commented:
We've added a new selector and refined some types in https://github.com/Automattic/wp-calypso/pull/61324. Shouldn't we remove that code now? Should we also update Titan to use this new approach?

daledupreez committed:
Remove unintended whitespace

daledupreez commented:
> We've added a new selector and refined some types in https://github.com/Automattic/wp-calypso/pull/61324. Shouldn't we remove that code now? Should we also update Titan to use this new approach?

I think we still need the selector that was added there, as it is answering a slightly different question. The selector I've added in this PR tracks whether an introductory offer is available for a product, while the selector in the other PR answers whether a particular domain is (in)eligible for an introductory offer (under the assumption that an introductory offer exists). Both selectors need to be true for us to show the introductory offer pricing, though for the upsell in the domain flow that isn't obvious because new domains are assumed to be eligible for the introductory offer.

As to whether we should update the Professional Email code to perform this check dynamically, I think we can do that, but I don't think we want to prioritise the work as we don't have any plans to change the free trials in the short term.
